### ISSUE TITLE
Problem with array string enumeration

### ISSUE THREAD

**Issue by Angelk90** [giddyinc/mongoose-to-swagger#8]
Hi @blugavere ,

I have a part of the schema of the model of this type, but the problem library correctly deciphers the "actions" part, he simply sees it as an array of objects,
in reality it is an array of enumerative strings.

ex:
```
{
...
action: ["create","delete",...]
...
}
```

Schema model:
```
scopes: [
      {
        scope: { type: String, enum: scopes },
        actions: [{ type: String, enum: actions }]
      }
    ]
```

Could something be done?

**Comment by blugavere** [comment]
fixed in `mongoose-to-swagger@0.1.0`

**Comment by Angelk90** [comment]
@blugavere : Thx.

**Comment by Angelk90** [comment]
@blugavere : 

It does not work properly I return something like this:

```
"actions": {
                "type": "array",
                "items": {
                  "type": "string"
                }
              }
```

I would like this to come back:

```
 "actions": {
                "type": "array",
                "items": {
                  "type": "string",
                  "enum": [
                    "create",
                    "read",
                    "update",
                    "delete"
                  ]
                }
              }
```

It also adds attributes that do not exist in the model, the attributes are as follows:

```
      "_id": {
          "type": "string"
        }
```
```
        "__v": {
          "type": "number"
        }
```

I solved the following problems in this way, but I don't think mine is a correct solution.


```
"use strict";

const adjustType = type => {
  if (type === undefined || type === null) return "Problem";
  let newType = type === "ObjectId" || type === "ObjectID" ? "object" : type;
  return newType.toLowerCase();
};

/*const returnElement = (ob, ...elements) => {
  const copy = Object.assign({}, ob);
  return elements.reduce((obj, key) => ({ ...obj, [key]: copy[key] }), {});
};

const print = mongooseProp => {
  let { instance, path, schema, caster } = returnElement(
    mongooseProp,
    "instance",
    "path",
    "schema",
    "caster"
  );

  console.log("---Instance:", instance, "");
  console.log("---Path:", path, "");
  if (caster) console.log("---Caster:", caster);
  if (schema)
    Object.keys(schema.paths)
      .map(x => schema.paths[x])
      .forEach(schemas => {
        console.log("---Schema:", schemas, "\n");
      });
  console.log("\n\n");
};*/

function documentModel(Model) {
  const obj = {
    title: Model.modelName,
    properties: {},
    required: []
  };

  const pathsToSchema = (parent, paths) => {
    Object.keys(paths)
      .map(x => paths[x])
      .forEach(mongooseProp => {
        //print(mongooseProp);
        if (
          mongooseProp &&
          mongooseProp.instance &&
          !(
            mongooseProp.instance === "ObjectID" && mongooseProp.path === "_id"
          ) &&
          !(mongooseProp.instance === "Number" && mongooseProp.path === "__v")
        ) {
          parent[mongooseProp.path] = {};
          const modelProp = parent[mongooseProp.path];

          if (mongooseProp.instance === "Array") {
            modelProp.type = "array";
            if (mongooseProp.schema) {
              modelProp.items = { properties: {} };
              pathsToSchema(
                modelProp.items.properties,
                mongooseProp.schema.paths
              );
            } else if (mongooseProp.caster) {
              modelProp.items = {};
              modelProp.type = "array";
              modelProp.items.type = "string";
              modelProp.items.enum = mongooseProp.caster.enumValues;
            } else {
              modelProp.items = { type: "object" };
            }
          } else if (mongooseProp.instance === "Embedded") {
            modelProp.properties = {};
            modelProp.type = "object";
            pathsToSchema(modelProp.properties, mongooseProp.schema.paths);
          } else if (mongooseProp.instance === "ObjectID") {
            modelProp.type = "string";
            modelProp.required = mongooseProp.isRequired || false;
          } else if (mongooseProp.instance === "Date") {
            modelProp.type = "string";
            modelProp.format = "date-time";
            modelProp.required = mongooseProp.isRequired || false;
          } else if (mongooseProp.instance !== undefined) {
            modelProp.type = adjustType(mongooseProp.instance);
            modelProp.required = mongooseProp.isRequired || false;
          }

          // custom mongoose options
          if (mongooseProp.options) {
            if (mongooseProp.options.description) {
              modelProp.description = mongooseProp.options.description;
            }
          }
          if (mongooseProp.enumValues && mongooseProp.enumValues.length) {
            modelProp.enum = mongooseProp.enumValues;
          }
          if (modelProp.required) {
            obj.required.push(mongooseProp.path);
          }
          delete modelProp.required;
        }
      });
  };

  pathsToSchema(obj.properties, Model.schema.paths);

  return obj;
}

documentModel.adjustType = adjustType;
module.exports = documentModel;
```


Model:
```
var mongoose = require("mongoose");

const CREATE = "create";
const READ = "read";
const UPDATE = "update";
const DELETE = "delete";
const actions = [CREATE, READ, UPDATE, DELETE];

const USER = "user";
const GEOAREA = "geo_area";
const APIKEY = "apikey";

const scopes = [USER, GEOAREA, APIKEY];

const apiKeySchema = new mongoose.Schema({
  Scopes: [
    {
      actions: [{ type: String, enum: actions }]
    }
  ]
});

const apiKeySchema2 = new mongoose.Schema(
  {
    owner: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
      description: "Owner user"
    },
    apiKey: String,
    expiration: Date,
    active: {
      type: Boolean,
      default: false
    },
    scopes: [
      {
        scope: { type: String, enum: scopes },
        actions: [{ type: String, enum: actions }]
      }
    ]
  },
  {
    timestamps: true
  }
);

const model = mongoose.model("ApiKey", apiKeySchema2);

module.exports = model;

```

**Comment by Angelk90** [comment]
@blugavere :

I tried to change your version, it seems to work at the moment let me know what you think.
I also managed to remove the attributes that it automatically added as `_id` and `__v`, I moved the part of the `required` over the `properties`, I think it is better to have this type of structure than the opposite, in which case you could do so that the user is given the possibility to choose the `required` above the `properties` or the opposite if you also want to keep the old structure.

```
/* eslint-disable eqeqeq */
/* eslint-disable no-eq-null */

"use strict";

const mapMongooseTypeToSwaggerType = type => {
  if (!type) {
    return null;
  }
  if (type === Number || type === "Number") {
    return "number";
  }
  if (type === String || type === "String") {
    return "string";
  }
  if (type === Boolean || type === "Boolean") {
    return "boolean";
  }
  if (type instanceof Function) {
    // special types
    if (type.name === "ObjectId" || type.name === "ObjectID") {
      return "string";
    }
    if (type.name === "Date") {
      return "string";
    }
    return type.name.toLowerCase();
  }

  if (type.type != null) {
    return mapMongooseTypeToSwaggerType(type.type);
  }

  if (Array.isArray(type)) {
    return "array";
  }

  return "object";
};

function documentModel(Model) {
  const obj = {
    title: Model.modelName,
    required: [],
    properties: {}
  };

  /*const keys = Object.keys(Model.schema.tree);
  console.log(keys);*/

  const pathsToSchema = (parent, paths) => {
    delete paths._id;
    delete paths.__v;
    Object.keys(paths)
      .map(x => paths[x])
      .forEach(mongooseProp => {
        parent[mongooseProp.path] = {};
        //const type = mapMongooseTypeToSwaggerType(mongooseProp);
        //console.log(type);
        const modelProp = parent[mongooseProp.path];

        if (mongooseProp.instance === "Array") {
          // console.log(mongooseProp.schema);
          modelProp.type = "array";
          modelProp.items = {
            properties: {}
          };

          if (mongooseProp.schema) {
            pathsToSchema(
              modelProp.items.properties,
              mongooseProp.schema.paths
            );
          } else if (
            mongooseProp.options &&
            mongooseProp.options.type != null &&
            (Array.isArray(mongooseProp.options.type) ||
              mongooseProp.options.type === "array")
          ) {
            modelProp.items = {
              type: mapMongooseTypeToSwaggerType(mongooseProp.options.type[0])
            };
            modelProp.items.enum = mongooseProp.caster.enumValues;
          } else {
            modelProp.items = {
              type: "object"
            };
          }
        } else if (mongooseProp.instance === "Embedded") {
          modelProp.properties = {};
          modelProp.type = "object";
          pathsToSchema(modelProp.properties, mongooseProp.schema.paths);
        } else if (mongooseProp.instance === "ObjectID") {
          modelProp.type = "string";
          modelProp.required = mongooseProp.isRequired || false;
        } else if (mongooseProp.instance === "Date") {
          modelProp.type = "string";
          modelProp.format = "date-time";
          modelProp.required = mongooseProp.isRequired || false;
        } else {
          modelProp.type = mapMongooseTypeToSwaggerType(mongooseProp.instance);
          modelProp.required = mongooseProp.isRequired || false;
        }
        // custom mongoose options
        if (mongooseProp.options) {
          if (mongooseProp.options.description) {
            modelProp.description = mongooseProp.options.description;
          }
        }
        if (mongooseProp.enumValues && mongooseProp.enumValues.length) {
          modelProp.enum = mongooseProp.enumValues;
        }
        if (modelProp.required) {
          obj.required.push(mongooseProp.path);
        }
        delete modelProp.required;
      });
  };

  pathsToSchema(obj.properties, Model.schema.paths);

  return obj;
}

documentModel.adjustType = mapMongooseTypeToSwaggerType;

module.exports = documentModel;

```

**Comment by Angelk90** [comment]
@blugavere : Are there any news?

**Comment by blugavere** [comment]
looking into an alternative implementation which relies on the 'tree' instead of the 'paths' for proper deep object support to fix all related issues

**Comment by blugavere** [comment]
pushed an update to a branch with the implementation which should cover most of your issues

still working out the `required` handling - you're welcome to take a look - think this was the only way to handle nested objects correctly

https://github.com/giddyinc/mongoose-to-swagger/pull/11

**Comment by Angelk90** [comment]
@blugavere : 

I'm having problems with .ts, you think you might be able to create an example on: https://codesandbox.io/s/node

So can we test it?

**Comment by blugavere** [comment]
fixed in 1.0

**Comment by Angelk90** [comment]
@blugavere : 
Could you try to put online a service where you can do tests?
